### DISCUSSION TITLE
[API Proposal]: add regex for pattern matching

### DISCUSSION THREAD

**Discussion by balakhaniyan** [dotnet/csharplang#8456] [Language Ideas]
### Background and motivation

Hello
Can dotnet team add a feature to pattern matching to use RegEx? (with `is` keyword)

### API Proposal

I think there is no need to submit an API proposal

### API Usage

```csharp
if (data is @"\s+") {

}
```

**Comment by huoyaoyuan** [comment]
This is a C# language feature and belongs to https://github.com/dotnet/csharplang/discussions.

Testing `IsMatch` itself will definitely not be a language feature. It's just a shorthand of one method `Regex.IsMatch`. This falls under *user-defined active pattern*, which is discussed at https://github.com/dotnet/csharplang/issues/277 and https://github.com/dotnet/csharplang/issues/4131 .
The more interesting case is to extract the matched captures directly as variables, like the opposite of string interpolation.

**Comment by Unknown6656** [comment]
The problem with your proposal is, that your example is already valid code:
https://sharplab.io/#v2:EYLgtghglgdgNAFxAJwK4wD4AEBMBGAWAChjis8AGAAgGcqBeKrLAIgB0aBqFgblKPIBOABR0odZuy4sAlHwF4RYiaw4BvPAF9ZPIA==

Meaning we cannot change how the existing expression of the type `... is "string"` behaves, as it would break existing code.

**Reply by balakhaniyan** [reply]
OK, but what about:
```C#
if (str is case @"\s+")
```
or 
```C#
if (str case @"\s+")
```
?

**Reply by Unknown6656** [reply]
That would indeed be a possibility, however, I am not sure whether the introduction of a new syntax is worth the effort. Additionally, how would you propose to implement the usage of [`RegexOptions`](https://learn.microsoft.com/en-us/dotnet/api/system.text.regularexpressions.regexoptions?view=net-8.0) in that case?

I, for example, nearly always use a combination of the following `RegexOptions` flags: `IgnoreCase`, `CultureInvariant`, `Compiled`, and `NonBacktracking`.

Should a syntax à la JavaScript be introduced?
```csharp
if (str is /\s+/i)
    ...
```

**Comment by jnm2** [comment]
If we got user-defined positional patterns (active patterns), we could help ourselves to regex patterns, ignore-case patterns, and anything we could dream of.

Currently championed at <https://github.com/dotnet/csharplang/issues/4131>.

**Reply by balakhaniyan** [reply]
And also some of them can be used in Expression Trees :), I know it has some challenges, but it is highly recommended

**Reply by jnm2** [reply]
Here's what expression trees are up against: https://github.com/dotnet/csharplang/discussions/4727#discussioncomment-4203498
